src/azul/plugins/repository/canned/__init__.py

```diff
@@ -134,6 +140,26 @@ def dss_deletion_subscription_query(self, prefix: str) -> JSON:
     def dss_subscription_query(self, prefix: str) -> JSON:
         return {}
 
+    def _construct_file_url(self, source_url: str, file_name: str) -> str:
+        """
+        >>> source_url = 'https://github.com/foo/bar/tree/de355cad/tests'
+        >>> Plugin._construct_file_url(Plugin, source_url, 'foo.zip')
+        'https://github.com/foo/bar/raw/de355cad/tests/foo.zip'
+
+        >>> Plugin._construct_file_url(Plugin, source_url, '')
+        Traceback (most recent call last):
+        ...
+        azul.RequirementError: file_name cannot be empty
```

Comment:
nit : might want to add a test for url encoding in the REF.

Reply:
Unfortunately with a URL structure `https://github.com/USER/REPO/tree/REF` and optional `/PATH`, if the `REF` piece contains `/` characters that are not URL encoded there is no way to determine which parts belong to the `REF` and which belong to the `PATH`.
`GitHubStagingAreaFactory.from_url` does the parsing and some validation of the URL, and any errors it couldn't determine would be revealed when `GitHubStagingAreaFactory.load_staging_area` attempts to read the files from the staging area.